[scroll-animations] Implement basic scroll timeline progress calculation

nmoucht opened the pull request:
#### 58bb30391e9051d43e0a2681f0df590ff88c9af9
<pre>
[scroll-animations] Implement basic scroll timeline progress calculation
<a href="https://bugs.webkit.org/show_bug.cgi?id=279800">https://bugs.webkit.org/show_bug.cgi?id=279800</a>
<a href="https://rdar.apple.com/136117479">rdar://136117479</a>

Reviewed by Simon Fraser.

Implement basic scroll timeline progress calculation based on the spec:
<a href="https://drafts.csswg.org/scroll-animations-1/#scroll-timeline-progress.">https://drafts.csswg.org/scroll-animations-1/#scroll-timeline-progress.</a>

* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeScrollTimelineData):
* Source/WebCore/animation/ScrollTimeline.h:

Canonical link: <a href="https://commits.webkit.org/284019@main">https://commits.webkit.org/284019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e44d6ac8cf6830fc2053c14ae9c0a2efea785f6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20756 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72160 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19242 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70214 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19058 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71164 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43477 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58841 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34880 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40146 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16255 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17599 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62114 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16601 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73856 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12068 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15875 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12107 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58914 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61883 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9797 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3409 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10373 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43290 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44364 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45558 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44105 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,35 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
     return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::No;
 }
 
+ScrollTimeline::ScrollTimelineData ScrollTimeline::computeScrollTimelineData()
```
Can the function be `const` ?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,35 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+    float scrollableOverflowSize;
+    float scrollContainerSize;
+
+    if (Ref { m_source->document() }->scrollingElement() == m_source.get()) {
```
This is repeated from the ViewTimeline PR. Let's share code.

nmoucht replied:
Currently I think the ViewTimeline code should call into this to get the data for the source element.

smfr commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -40,6 +40,12 @@ class RenderStyle;
 class ScrollTimeline : public AnimationTimeline {
 public:
+    struct ScrollTimelineData {
+        float scrollableOverflowSize;
```
Initialize to zero.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -40,6 +40,12 @@ class RenderStyle;
 
 class ScrollTimeline : public AnimationTimeline {
 public:
+    struct ScrollTimelineData {
```
You can make this `private` and rename to just `Data`.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,39 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+
+    float scrollContainerSize = axis() == ScrollAxis::Block ? sourceScrollableArea->visibleHeight() : sourceScrollableArea->visibleWidth();
+    float scrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->scrollPosition().y() : sourceScrollableArea->scrollPosition().x();
+    float scrollableOverflowSize = m_axis == ScrollAxis::Block ? renderBox->scrollHeight() : renderBox->scrollWidth();
```
You can use `ScrollableArea.contentsSize()` so that this line uses `sourceScrollableArea` like the lines above.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,39 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+    return { scrollableOverflowSize , scrollContainerSize, scrollOffset };
+}
+
+std::optional<Seconds> ScrollTimeline::currentTime()
```
Are we going to plumb through some "animation value" type that isn't Seconds, to avoid future confusion?

nmoucht replied:
Yeah Antoine is currently working on changing everything to use a type that works with Seconds and percentage types. For now I have it like this to make it easier for me to test this code.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,39 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+std::optional<Seconds> ScrollTimeline::currentTime()
+{
+    auto data = computeScrollTimelineData();
+    return Seconds { data.scrollOffset / (data.scrollableOverflowSize - data.scrollContainerSize) };
```
`data.scrollableOverflowSize - data.scrollContainerSize` is exactly `ScrollableArea::maximumScrollOffset()` so you should use that from the get-go.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -31,6 +31,8 @@
 #include "CSSScrollValue.h"
 #include "CSSValuePool.h"
 #include "Element.h"
+#include "RenderLayerScrollableArea.h"
```
Do you need either of these includes now?

nmoucht replied:
These are still needed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,39 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
 
+ScrollableArea* ScrollTimeline::scrollableAreaForSourceRenderer(CheckedPtr<RenderBox> renderer, Ref<Document> document)
+{
+    return renderer->element() == document->documentElement() ? (ScrollableArea*)&renderer->view().frameView() : (ScrollableArea*)renderer->layer()->scrollableArea();
```
I don't think you need either of the `(ScrollableArea*)` casts.

smfr replied:
Also you are not checking `renderer->hasLayer()`, so `renderer->layer()` will often crash. Don't you need to find the renderer's enclosing ScrollableArea somewhere?

nmoucht replied:
Right added a check for layer() and canBeScrolledAndHasScrollableArea(). We don't need to look for the enclosing scrollable area, that is just for ViewTimelines. ScrollTimelines are provided the source element when created through the api.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,40 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+        return nullptr;
+
+    if (renderer->element() == document->documentElement())
+        return &renderer->view().frameView();
```
I prefer blank lines after returns (be consistent with the one above).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,40 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+
+    if (renderer->element() == document->documentElement())
+        return &renderer->view().frameView();
+    return (renderBox->canBeScrolledAndHasScrollableArea() && renderBox->layer()) ? renderBox->layer()->scrollableArea() : nullptr;
```
Nit: `box->hasLayer()` is a bit check, and possibly slightly faster than checking `box->layer()`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,40 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+
+    auto* sourceScrollableArea = scrollableAreaForSourceRenderer(m_source->renderer(), m_source->document());
+    if (!sourceScrollableArea)
+        return { };
```
Should this assert? Can we get here with a ScrollTimeline having a renderer that is not scrollable?

nmoucht replied:
Its incorrect to give a non-scrollable element as a source but we shouldn't assert.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -138,4 +140,40 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+    // scroll offset ÷ (scrollable overflow size − scroll container size)
+
+    float maxScrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->maximumScrollOffset().y() : sourceScrollableArea->maximumScrollOffset().x();
+    float scrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->scrollPosition().y() : sourceScrollableArea->scrollPosition().x();
```
Beware of `scrollPosition` vs `scrollOffset`, and don't mix and match. They differ in RTL horizontal scrollers, so you need some tests for those. But use `scrollOffset()` here.

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Implement basic scroll timeline progress calculation
https://bugs.webkit.org/show_bug.cgi?id=279800
rdar://136117479

Reviewed by Simon Fraser.

Implement basic scroll timeline progress calculation based on the spec:
https://drafts.csswg.org/scroll-animations-1/#scroll-timeline-progress.

* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeScrollTimelineData):
* Source/WebCore/animation/ScrollTimeline.h:

Canonical link: https://commits.webkit.org/284019@main

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -140,4 +142,35 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
+    if (!m_source)
+        return { };
+
+    auto* sourceScrollableArea = scrollableAreaForSourceRenderer(m_source->renderer(), m_source->document());
```
Should this have been a `CheckedPtr`?

graouts commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -58,11 +59,19 @@ class ScrollTimeline : public AnimationTimeline {
 
 private:
+    struct Data {
+        float maxScrollOffset = 0;
+        float scrollOffset = 0;
```
You should consider adopting the `{ 0 }` style here.
